File: Vehicle.py
from matplotlib import pyplot as plt
import ggv.GGV as GGV
import setups
import state_models
import trajectory.Trajectory as Trajectory
from numpy.polynomial import Polynomial
import numpy as np
import math


class Vehicle:
    def __init__(self, trajectory_path, params, is_closed, mesh_resolution=10):
        self.params = params

        '''Total Reduction'''
        self.gear_tot = self.params.gear_ratios[-1] * self.params.final_drive * self.params.primary_reduction
        '''Max Achievable Velocity, assuming RPM never exceeds shiftpoint'''
        self.v_max = self.params.shiftpoint / (self.gear_tot/self.params.tire_radius*60/(2 * math.pi))
        self.GGV = GGV.GGV(self.params, self.gear_tot, self.v_max)
        
        

        self.trajectory = Trajectory.Trajectory(trajectory_path, is_closed, self.GGV.radii_range[0], self.GGV.radii_range[-1])
        self._interval = mesh_resolution

        self._mesh_size = (self.trajectory.num_points - 1) * self._interval
        self.count = np.zeros(self._mesh_size)

        self.time = np.zeros(self._mesh_size)

        self.x = np.zeros(self._mesh_size)
        self.y = np.zeros(self._mesh_size)

        '''Vehicle State Attributes'''
        self.dist = np.zeros(self._mesh_size)
        self.dist_f = np.zeros(self._mesh_size)
        self.dist_r = np.zeros(self._mesh_size)

        self.turn_dir = np.zeros(self._mesh_size)

        self.velocity = np.zeros(self._mesh_size)
        self.velocity_f = np.zeros(self._mesh_size)
        self.velocity_r = np.zeros(self._mesh_size)

        self.gear = np.zeros(self._mesh_size)
        self.is_shifting = np.zeros(self._mesh_size)

        self.ax = np.zeros(self._mesh_size)
        self.ax_f = np.zeros(self._mesh_size)
        self.ax_r = np.zeros(self._mesh_size)

        self.ay = np.zeros(self._mesh_size)
        self.ay_f = np.zeros(self._mesh_size)
        self.ay_r = np.zeros(self._mesh_size)
        self.fz_fr = []
        self.fz_fl = []
        self.fz_rr = []
        self.fz_rl = []
        self.fz_total = []
        self.cgz = []

    
        def __del__(self):
            self._matlab_engine.quit()


    def reset_ggv(self):
        self.GGV = GGV.GGV(self.AERO, self.DYN, self.PTN, self.params, self.gear_tot, self.v_max, self._matlab_engine)

    def simulate_accel(self):
        vel = 1e-6
        gear = 1
        required_gear = gear
        time_shifting = 0
        is_shifting = False
        count = 0
        distance= 0
        time = 0

        for d in range(75):
            #x_1 = self.trajectory.points[0][point_idx]
            #x_2 = self.trajectory.points[0][(point_idx + 1) % self.trajectory.num_points]

            '''Distance of trajectory interval in meters'''
            dist = 1

            for idx, v in enumerate(self.GGV.velocity_range):
                if(v > vel):
                    required_gear = self.GGV.expected_gears[idx]
                    break
            
            is_shifting = required_gear != gear

            AX_cap = self.GGV.accel_capability.evaluate(vel)
            delta_d = dist / self._interval

            for j in range(self._interval):
                self.x[count] = d + dist * j/self._interval
                self.y[count] = 0 * j/self._interval

                self.ay[count] = 0

                self.turn_dir[count] = 1

                self.count[count] = count
                self.velocity[count] = vel
                self.gear[count] = gear
                self.dist[count] =  distance + delta_d * j

                v_max = self.v_max - 1

                dt = delta_d / vel
                self.time[count] = time + dt * j
                if is_shifting and vel < v_max:
                    # Currently shifting, do not accelerate
                    dt = delta_d / vel
                    time_shifting += dt
                    self.ax[count] = 0
                    self.is_shifting[count] = 1
                elif vel < v_max:
                    # Take a slice of the GGV using AX_cap and AY_cap as verticies
                    # the actual lateral acceleration will define the remaining longitudinal acceleration.
                    self.ax[count] = AX_cap

                    # Solve the 1D kinematic equation for time
                    p = Polynomial([-delta_d, vel, 0.5*9.81*AX_cap])
                    tt = p.roots()
                    dt = max(tt)

                    #Solve for change in velocity
                    dv = 9.81 * AX_cap * dt

                    #Clamp new velocity to v_max
                    vel = min(v_max, vel + dv)

                    for i in range(len(self.GGV.expected_gears)):
                        if(self.GGV.expected_gears[i] > vel):
                            required_gear = self.GGV.expected_gears[i]
                    
                    is_shifting = required_gear > gear
                else:
                    # Vehicle is neither shifting or accelerating, is at vmax
                    vel = v_max
                    dt = delta_d / vel
                    self.ax[count] = 0

                

                if time_shifting >= self.params.shift_time:
                    is_shifting = False
                    time_shifting = 0
                    gear = required_gear
                
                count += 1
                time += dt
            
            distance += dist

        return max(self.time)

    def simulate_endurance(self):
        vel = 20 * 0.3048; # m/s
        self.simulate_forwards(vel)
        self.simulate_reverse()
        self.simulate_forwards(self.velocity_r[0])

        start_vel = []
        end_vel = []
        in_brake = False

        for i in self.count:
            i = int(i)
            self.dist[i] = self.dist_f[i]
            if self.velocity_f[i] < self.velocity_r[i]:
                self.velocity[i] = self.velocity_f[i]
                self.ax[i] = self.ax_f[i]
                self.ay[i] = self.ay_f[i]
            else:
                self.velocity[i] = self.velocity_r[i]
                self.ax[i] = -1 * self.ax_r[i]
                self.ay[i] = self.ay_r[i]
            
            '''Snippet for Little Liam I think, some brakes stuff idr'''
            if(i > 0):
                if self.ax[i] < 0 and self.ax[i-1] > 0:
                    start_vel.append(self.velocity[i-1])
                    in_brake = True
                elif (in_brake) and (self.ax[i] > 0 and self.ax[i-1] <= 0):
                    end_vel.append(self.velocity[i-1])
                    in_brake = False
        

        '''Determine turn handedness and remove outliers'''
        self.ay[self.ay > self.GGV.lateral_capability.evaluate(self.v_max + 1)] = self.GGV.lateral_capability.evaluate(self.v_max + 1)
        self.ay_f[self.ay_f > self.GGV.lateral_capability.evaluate(self.v_max + 1)] = self.GGV.lateral_capability.evaluate(self.v_max + 1)
        self.ay_r[self.ay_r > self.GGV.lateral_capability.evaluate(self.v_max + 1)] = self.GGV.lateral_capability.evaluate(self.v_max + 1)
        
        self.ay = np.multiply(self.ay, self.turn_dir)
        self.ay_f = np.multiply(self.ay_f, self.turn_dir)
        self.ay_r = np.multiply(self.ay_r, self.turn_dir)
        
        self.ax_r = np.multiply(self.ax_r, -1)
        
        
        # self.ax.plot(self.dist_f, math.comb)
        # plt.show()
        # plt.scatter(self.x, self.y, c=math.comb)
        # plt.colorbar()
        # plt.show()
        # fig, ax = plt.subplots()
        # ax.scatter(self.x,self.y,marker=',')
        # for i, txt in enumerate(range(len(self.x))):
        #     ax.annotate(txt,(self.x[i],self.y[i]))

        print('POINTS',len(self.velocity))

        # Gathering data for plots
        self.cgz=[]
        self.vtest=[]
        self.roll=[]
        self.pitch=[]
        self.axp=[]
        self.ayp=[]
        self.xpos = []
        self.ypos = []

        for i in range(self.trajectory.num_points - 1) :
            i = int(i)
            r = self.trajectory.radii[i]
            state_in = state_models.StateInput(Ax=self.ax[i], Ay=self.ay[i], v=self.velocity[i], r=r, delta=0, beta=0)
            # state_in = state_models.StateInput(Ax=0, Ay=0, v=self.velocity[i], r=r, delta=0, beta=0) # Constant accel variation
            self.setup = setups.Panda
            v =state_models.VehicleState(params=self.params)
            v.eval(state_in=state_in)
            # self.fz_rr.append(v.rr_tire.Fz)
            # self.fz_rl.append(v.rl_tire.Fz)
            # self.fz_fr.append(v.fr_tire.Fz)
            # self.fz_fl.append(v.fl_tire.Fz)
            # self.fz_total.append(v.rr_tire.Fz)
            # self.fz_total.append(v.rl_tire.Fz)
            # self.fz_total.append(v.fr_tire.Fz)
            # self.fz_total.append(v.fl_tire.Fz)
            self.cgz.append(v.cgz)
            self.vtest.append(v.v)
            self.roll.append(v.phi)
            self.pitch.append(v.theta)
            self.axp.append(v.Ax)
            self.ayp.append(v.Ay)
        return max(self.time)  
    
    def simulate_forwards(self, starting_v):
        vel = starting_v
        gear = 1
        required_gear = gear
        time_shifting = 0
        is_shifting = False
        count = 0
        distance= 0
        time = 0
        for point_idx in (range(self.trajectory.num_points - 1)):
            x_1 = self.trajectory.points[0][point_idx]
            y_1 = self.trajectory.points[1][point_idx]
            x_2 = self.trajectory.points[0][(point_idx + 1)]
            y_2 = self.trajectory.points[1][(point_idx + 1)]
            '''Distance of trajectory interval in meters'''
            dist = math.sqrt((x_1-x_2)**2 + (y_2-y_1)**2)

            r = self.trajectory.radii[point_idx]

            '''Max Achievable Cornering Velocity through this segment'''
            v_max = min(self.v_max, self.GGV.cornering_capability.evaluate(r))

            for idx, v in enumerate(self.GGV.velocity_range):
                if(v > vel):
                    required_gear = self.GGV.expected_gears[idx]
                    break
            
            is_shifting = required_gear != gear

            AX_cap = self.GGV.accel_capability.evaluate(vel)
            AY_cap = self.GGV.lateral_capability.evaluate(vel)
            AY_actual = vel ** 2 / (r * 9.81)

            delta_d = dist / self._interval

            for j in range(self._interval):
                self.x[count] = x_1 + (x_2 - x_1) * j/self._interval
                self.y[count] = y_1 + (y_2 - y_1) * j/self._interval

                self.ay_f[count] = min(AY_cap,AY_actual)

                self.turn_dir[count] = np.sign(self.trajectory._curvature[point_idx])

                self.count[count] = count
                self.velocity_f[count] = vel
                self.gear[count] = gear
                self.dist_f[count] =  distance + delta_d * j
 
                dt = delta_d / vel
                self.time[count] = time + dt * j
                if is_shifting and vel < v_max:
                    # Currently shifting, do not accelerate
                    dt = delta_d / vel
                    time_shifting += dt
                    self.ax_f[count] = 0
                    self.is_shifting[count] = 1
                elif vel < v_max:
                    # Take a slice of the GGV using AX_cap and AY_cap as verticies
                    # the actual lateral acceleration will define the remaining longitudinal acceleration.
                    ax_potential = AX_cap*math.sqrt(1-(min(AY_cap,AY_actual)/AY_cap)**2)
                    self.ax_f[count] = ax_potential

                    # Solve the 1D kinematic equation for time
                    p = Polynomial([-delta_d, vel, 0.5*9.81*ax_potential])
                    tt = p.roots()
                    dt = max(tt)

                    #Solve for change in velocity
                    dv = 9.81 * ax_potential * dt

                    #Clamp new velocity to v_max
                    vel = min(v_max, vel + dv)

                    for i in range(len(self.GGV.expected_gears)):
                        if(self.GGV.expected_gears[i] > vel):
                            required_gear = self.GGV.expected_gears[i]
                    
                    is_shifting = required_gear > gear
                else:
                    # Vehicle is neither shifting or accelerating, is at vmax
                    vel = v_max
                    dt = delta_d / vel
                    self.ax_f[count] = 0

                

                if time_shifting >= self.params.shift_time:
                    is_shifting = False
                    time_shifting = 0
                    gear = required_gear
                
                count += 1
                time += dt

            distance += dist

    def simulate_reverse(self):
        vel = self.velocity_f[-1]
        distance = 0
        count = int(self.count[-1])
        for point_idx in range(self.trajectory.num_points - 1, -1, -1):
            x_1 = self.trajectory.points[0][point_idx]
            y_1 = self.trajectory.points[1][point_idx]
            x_2 = self.trajectory.points[0][(point_idx - 1)]
            y_2 = self.trajectory.points[1][(point_idx - 1)]

            '''Distance of trajectory interval in meters'''
            dist = math.sqrt((x_1-x_2)**2 + (y_2-y_1)**2)

            r = self.trajectory.radii[point_idx]

            '''Max Achievable Cornering Velocity through this segment'''
            v_max = min(self.v_max, self.GGV.cornering_capability.evaluate(r))

            AX_cap = -1 * self.GGV.braking_capability.evaluate(vel)
            AY_cap = self.GGV.lateral_capability.evaluate(vel)
            AY_actual = vel ** 2 / (r * 9.81)

            delta_d = dist/self._interval
            for j in range(self._interval):
                self.velocity_r[count] = vel
                self.dist_r[count] =  distance + delta_d * j

                self.ay_r[count] = min(AY_cap,AY_actual)
                if vel < v_max:
                    AX_actual = AX_cap*(1-(min(AY_cap,AY_actual)/AY_cap)**2)
                    self.ax_r[count] = AX_actual
                    p = Polynomial([-delta_d, vel, 0.5*9.81*AX_actual])
                    tt = p.roots()
                    dt = max(tt)
                    dv = 9.81 * AX_actual * dt
                    dv_max = v_max - vel
                    vel += min(dv, dv_max)
                else:
                    vel = v_max
                    self.ax_r[count] = 0

                count -= 1
            distance += dist

File: julia/src/Vehicle.jl
mutable struct Vehicle
    params::Params
    mf52::MF52
    gear_tot::Float64
    v_max::Float64
    ggv::GGV
    trajectory::Trajectory
    track_path::Union{TrackPath,Nothing}  # retained for future path queries; nothing when loaded from CSV
    _interval::Int
    _mesh_size::Int
    # Mesh arrays
    count::Vector{Float64}
    time::Vector{Float64}
    x::Vector{Float64}; y::Vector{Float64}
    dist::Vector{Float64}; dist_f::Vector{Float64}; dist_r::Vector{Float64}
    turn_dir::Vector{Float64}
    velocity::Vector{Float64}; velocity_f::Vector{Float64}; velocity_r::Vector{Float64}
    gear::Vector{Float64}; is_shifting::Vector{Float64}
    ax::Vector{Float64}; ax_f::Vector{Float64}; ax_r::Vector{Float64}
    ay::Vector{Float64}; ay_f::Vector{Float64}; ay_r::Vector{Float64}
    # Post-processed
    cgz::Vector{Float64}; roll::Vector{Float64}; pitch::Vector{Float64}
end

"""
    Vehicle(params, track; base_dir, calc_lateral, mesh_resolution)

Construct a `Vehicle` from a `TrackPath`. The `TrackPath` is converted to a
`Trajectory` internally so the lap simulation runs unchanged. The original
`TrackPath` is stored on the vehicle for future use (track-width queries, etc.).
"""
function _vehicle_from_parts(params::Params, mf52::MF52, traj::Trajectory;
                             calc_lateral::Bool=true, mesh_resolution::Int=10)
    gear_tot = params.gear_ratios[end] * params.final_drive * params.primary_reduction
    v_max    = params.shiftpoint / (gear_tot/params.tire_radius * 60.0/(2π))
    ggv      = GGV(params, mf52, gear_tot, v_max; calc_lateral=calc_lateral)
    generate!(ggv)
    iv       = mesh_resolution
    msz      = (traj.num_points - 1) * iv
    z()      = zeros(msz)
    Vehicle(params, mf52, gear_tot, v_max, ggv, traj, nothing, iv, msz,
            z(),z(),z(),z(), z(),z(),z(), z(),
            z(),z(),z(), z(),z(), z(),z(),z(), z(),z(),z(),
            Float64[], Float64[], Float64[])
end

function Vehicle(params::Params, trajectory_path::String, is_closed::Bool;
                 base_dir::String=".", calc_lateral::Bool=true, mesh_resolution::Int=10)
    mf52 = MF52(base_dir)
    traj = Trajectory(trajectory_path, is_closed, 3.5, 36.0)
    return _vehicle_from_parts(params, mf52, traj;
                               calc_lateral=calc_lateral, mesh_resolution=mesh_resolution)
end

function Vehicle(params::Params, mf52::MF52, traj::Trajectory;
                 calc_lateral::Bool=true, mesh_resolution::Int=10)
    return _vehicle_from_parts(params, mf52, traj;
                               calc_lateral=calc_lateral, mesh_resolution=mesh_resolution)
end

# ── Kinematic helper ─────────────────────────────────────────────
function solve_dt(delta_d::Float64, vel::Float64, ax::Float64)::Float64
    a_c = 0.5*9.81*ax; b_c = vel; c_c = -delta_d
    abs(a_c) < 1e-12 && return delta_d/vel
    disc = b_c^2 - 4.0*a_c*c_c
    disc < 0.0 && return delta_d/vel
    sq = sqrt(disc)
    t1 = (-b_c+sq)/(2.0*a_c); t2 = (-b_c-sq)/(2.0*a_c)
    cands = filter(t->t>0.0, [t1,t2])
    isempty(cands) && return delta_d/vel
    return maximum(cands)
end

# ── Forward pass ─────────────────────────────────────────────────
function simulate_forwards!(v::Vehicle, starting_vel::Float64)
    vel=starting_vel; gear=1; req_gear=1
    t_shift=0.0; shifting=false; cnt=0; dist_acc=0.0; time=0.0

    for pi in 1:(v.trajectory.num_points-1)
        x1=v.trajectory.points[1][pi]; y1=v.trajectory.points[2][pi]
        x2=v.trajectory.points[1][pi+1]; y2=v.trajectory.points[2][pi+1]
        seg = sqrt((x1-x2)^2+(y1-y2)^2)
        r   = v.trajectory.radii[pi]
        v_max = min(v.v_max, evaluate(v.ggv.cornering_capability, r))

        for (ei,vv) in enumerate(v.ggv.velocity_range)
            if Float64(vv) > vel; req_gear = v.ggv.expected_gears[ei]; break; end
        end
        shifting = (req_gear != gear)

        AX_cap  = evaluate(v.ggv.accel_capability,  vel)
        AY_cap  = evaluate(v.ggv.lateral_capability, vel)
        AY_act  = vel^2/(r*9.81)
        ddelta  = seg/v._interval

        for j in 0:(v._interval-1)
            idx = cnt+1
            v.x[idx]=x1+(x2-x1)*j/v._interval; v.y[idx]=y1+(y2-y1)*j/v._interval
            v.ay_f[idx] = min(AY_cap, AY_act)
            v.turn_dir[idx] = sign(v.trajectory._curvature[pi])
            v.count[idx]    = Float64(cnt)
            v.velocity_f[idx] = vel
            v.gear[idx]     = Float64(gear)
            v.dist_f[idx]   = dist_acc + ddelta*j
            dt = ddelta/vel
            v.time[idx] = time + dt*j

            if shifting && vel < v_max
                t_shift += ddelta/vel; v.ax_f[idx]=0.0; v.is_shifting[idx]=1.0
            elseif vel < v_max
                ax_pot = AX_cap * sqrt(1.0-(min(AY_cap,AY_act)/AY_cap)^2)
                v.ax_f[idx] = ax_pot
                dt  = solve_dt(ddelta, vel, ax_pot)
                vel = min(v_max, vel + 9.81*ax_pot*dt)
                for i in eachindex(v.ggv.expected_gears)
                    Float64(v.ggv.expected_gears[i]) > vel && (req_gear=v.ggv.expected_gears[i])
                end
                shifting = (req_gear > gear)
            else
                vel=v_max; dt=ddelta/vel; v.ax_f[idx]=0.0
            end

            t_shift >= v.params.shift_time && (shifting=false; t_shift=0.0; gear=req_gear)
            cnt+=1; time+=dt
        end
        dist_acc+=seg
    end
end

# ── Reverse pass ─────────────────────────────────────────────────
function simulate_reverse!(v::Vehicle)
    vel=v.velocity_f[end]; dist_acc=0.0
    cnt=Int(v.count[end])                        # 0-based index of last point

    for pi in v.trajectory.num_points:-1:1
        x1=v.trajectory.points[1][pi];  y1=v.trajectory.points[2][pi]
        # wrap-around for closed track: index 0 (Python) → index n (Julia, last)
        pi_prev = pi==1 ? v.trajectory.num_points : pi-1
        x2=v.trajectory.points[1][pi_prev]; y2=v.trajectory.points[2][pi_prev]
        seg = sqrt((x1-x2)^2+(y1-y2)^2)
        r   = v.trajectory.radii[pi]
        v_max = min(v.v_max, evaluate(v.ggv.cornering_capability, r))

        AX_cap  = -1.0*evaluate(v.ggv.braking_capability, vel)
        AY_cap  = evaluate(v.ggv.lateral_capability, vel)
        AY_act  = vel^2/(r*9.81)
        ddelta  = seg/v._interval

        for j in 0:(v._interval-1)
            # Use modular indexing for the extra wrap segment
            idx = mod(cnt, v._mesh_size) + 1    # 1-based, wraps
            v.velocity_r[idx] = vel
            v.dist_r[idx]     = dist_acc + ddelta*j
            v.ay_r[idx]       = min(AY_cap, AY_act)

            if vel < v_max
                AX_act = AX_cap*(1.0-(min(AY_cap,AY_act)/AY_cap)^2)
                v.ax_r[idx] = AX_act
                dt  = solve_dt(ddelta, vel, AX_act)
                dv  = 9.81*AX_act*dt
                vel += min(dv, v_max-vel)
            else
                vel=v_max; v.ax_r[idx]=0.0
            end
            cnt-=1
        end
        dist_acc+=seg
    end
end

# ── Endurance merge ───────────────────────────────────────────────
function simulate_endurance!(v::Vehicle)::Float64
    simulate_forwards!(v, 20.0*0.3048)
    simulate_reverse!(v)
    simulate_forwards!(v, v.velocity_r[1])

    ay_max = evaluate(v.ggv.lateral_capability, v.v_max+1.0)
    for i in eachindex(v.count)
        v.dist[i] = v.dist_f[i]
        if v.velocity_f[i] < v.velocity_r[i]
            v.velocity[i]=v.velocity_f[i]; v.ax[i]=v.ax_f[i]; v.ay[i]=v.ay_f[i]
        else
            v.velocity[i]=v.velocity_r[i]; v.ax[i]=-v.ax_r[i]; v.ay[i]=v.ay_r[i]
        end
    end

    for i in eachindex(v.ay)
        v.ay[i]   > ay_max && (v.ay[i]=ay_max)
        v.ay_f[i] > ay_max && (v.ay_f[i]=ay_max)
        v.ay_r[i] > ay_max && (v.ay_r[i]=ay_max)
    end
    v.ay   .*= v.turn_dir; v.ay_f .*= v.turn_dir; v.ay_r .*= v.turn_dir
    v.ax_r .*= -1.0

    # Post-process vehicle state per trajectory point
    for i in 1:(v.trajectory.num_points-1)
        r  = v.trajectory.radii[i]
        si = StateInput(Ax=v.ax[i], Ay=v.ay[i], v=v.velocity[i], r=r)
        vs = VehicleState(v.mf52, v.params)
        eval!(vs, si)
        push!(v.cgz,  vs.cgz)
        push!(v.roll,  vs.phi)
        push!(v.pitch, vs.theta)
    end

    return maximum(v.time)
end
